pkg/mountexecutor/mountexecutor.go

```diff
@@ -0,0 +1,127 @@
+package mountexecutor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/containerd/containerd/reference/docker"
+	"github.com/warm-metal/csi-driver-image/pkg/backend"
+	"github.com/warm-metal/csi-driver-image/pkg/mountstatus"
+	"github.com/warm-metal/csi-driver-image/pkg/pullstatus"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+const (
+	mountPollTimeInterval = 100 * time.Millisecond
+	mountPollTimeout      = 2 * time.Minute
+	mountCtxTimeout       = 10 * time.Minute
+)
+
+// MountExecutorOptions are options passed to mount executor
+type MountExecutorOptions struct {
+	AsyncMount bool
+	Mounter    *backend.SnapshotMounter
+}
+
+// MountOptions are options for a single mount request
+type MountOptions struct {
+	// Context here is only valid for synchronous mounts
+	Context          context.Context
+	NamedRef         docker.Named
+	VolumeId         string
+	TargetPath       string
+	VolumeCapability *csi.VolumeCapability
+	ReadOnly         bool
+}
+
+// MountExecutor executes mount
+type MountExecutor struct {
+	asyncMount bool
+	mutex      *sync.Mutex
+	mounter    *backend.SnapshotMounter
+	asyncErrs  map[docker.Named]error
+}
+
+// NewMountExecutor initializes a new mount executor
+func NewMountExecutor(o *MountExecutorOptions) *MountExecutor {
+	return &MountExecutor{
+		asyncMount: o.AsyncMount,
+		mutex:      &sync.Mutex{},
+		mounter:    o.Mounter,
+	}
+}
+
+// StartMounting starts the mounting
+func (m *MountExecutor) StartMounting(o *MountOptions) error {
+
+	if pullstatus.Get(o.NamedRef) != pullstatus.Pulled || mountstatus.Get(o.NamedRef) == mountstatus.StillMounting {
+		return nil
```

Comment:
Can we log this so that we would know what happened to various mount requests incoming from kubelet ?

Reply:
I will add the log 👍